### PR TITLE
ROOT I/O bug fixes and improvements

### PR DESCRIPTION
**Pull request by amadio**
Fix nullptr dereference bugs in TClass and TList, wrap TClass.cxx to 120 columns to follow coding conventions, and avoid StreamerInfo lookup when it is cached.

### CODE REVIEW

**Comment by amadio** [comment]
Note: reformatted with clang-format and pushed again.

File: core/meta/inc/TClass.h [hunk 1]
```diff
@@ -544,8 +553,16 @@ friend class TProtoClass;
 };
 
 namespace ROOT {
-   template <typename T> TClass* GetClass(      T* /* dummy */)        { return TClass::GetClass(typeid(T)); }
-   template <typename T> TClass* GetClass(const T* /* dummy */)        { return TClass::GetClass(typeid(T)); }
+template <typename T>
```
Comment (Axel-Naumann) [inline core/meta/inc/TClass.h]: Is this really worth changing? I find the original version much more obvious to read.

File: core/cont/src/TList.cxx [hunk 2]
```diff
@@ -713,30 +716,31 @@ void TList::RecursiveRemove(TObject *obj)
...
-               lnk->Prev()->fNext = next;
-               lnk->Next()->fPrev = lnk->Prev();
-               DeleteLink(lnk);
-            }
-            fSize--;
-            fCache = 0;
-            Changed();
-         } else
-            ob->RecursiveRemove(obj);
+      if (TObject *ob = lnk->GetObject()) {
```
Comment (Axel-Naumann) [inline core/cont/src/TList.cxx]: Do you think this can be converted into a `for` loop, too? Maybe even with
```
if (!ob->TestBit(kNotDeleted))
   continue;
if (!ob->IsEqual(obj)) {
   ob->RecursiveRemove(obj);
   continue;
}
```
to reduce brain stack and when reading this code, and indentation?

File: core/meta/src/TClass.cxx [hunk 3]
```diff
@@ -4359,29 +4359,33 @@ Int_t TClass::GetNmethods()
...
+
+   // Warning: version may be -1 for an emulated class, or -2 if the
+   //          user requested the emulated streamerInfo for an abstract
+   //          base class, even though we have a dictionary for it.
+
+   if ((version < -1) || (version >= fStreamerInfo->GetSize())) {
       Error("GetStreamerInfo", "class: %s, attempting to access a wrong version: %d", GetName(), version);
       // FIXME: Shouldn't we go to -1 here, or better just abort?
-      version = 0;
+      version = fClassVersion;
```
Comment (Axel-Naumann) [inline core/meta/src/TClass.cxx]: That change is a functional one, right? I don't see it motivated by the commit message ("Avoid lookup for StreamerInfo when it is cached"); could you maybe even move this into a separate commit explaining why this change should happen? (I am not questioning that it's a good thing, I bet you discussed with Philippe, but these changes are not obvious to pinpoint when looking at commit logs.)

File: core/meta/inc/TClass.h (hunk 1 above)
Comment (amadio) [reply core/meta/inc/TClass.h]: This was clang-format doing its job...

File: core/cont/src/TList.cxx (hunk 2 above)
Comment (amadio) [reply core/cont/src/TList.cxx]: Sure, I will try.

File: core/meta/src/TClass.cxx (hunk 3 above)
Comment (amadio) [reply core/meta/src/TClass.cxx]: I need @pcanal  to take a look, actually, that's why I included this commit again. Let's see what he says. If `version = 0` is used, then it will check for the `StreamerInfo` every time. The version itself is cached in `fClassVersion`, so you can only avoid the lookup when it's already there if you check against the same number. `version = 0` means current, which is whatever is stored in `fClassVersion`.

**Comment by amadio** [comment]
The test failure is unrelated to this PR.

File: core/meta/src/TClass.cxx (hunk 3 above)
Comment (pcanal) [reply core/meta/src/TClass.cxx]: > version = 0 means current, 

Technically, no anymore at this point in the code.  It now means literally the one in slot zero .. but there is never any StreamerInfo in that slot, so it seems it leads to the execution of line 4396 and hence the leak of a TStreamerInfo (in cases where the fClassVersion's StreamerInfo was already build.

This code snippet is handling the error case (an non-existing StreamerInfo was asked for) and the new value of 'version' was essentially arbitrary and Guilherme's value is an improvement.

File: core/meta/src/TClass.cxx [hunk 4]
```diff
@@ -4359,29 +4359,33 @@ Int_t TClass::GetNmethods()
...
-   if (version == 0) {
+   if (version == 0)
       version = fClassVersion;
-   }
-   Int_t ninfos = fStreamerInfo->GetSize();
-   if ((version < -1) || (version >= ninfos)) {
+
+   // If the StreamerInfo is assigned to the fLastReadInfo, we are
+   // guaranteed it was built and compiled.
+   if (sinfo && sinfo->GetClassVersion() == version)
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: Reminder of the discussion on this line, in the previous PR

Comment (pcanal) [reply core/meta/src/TClass.cxx]: >  	
> @Axel-Naumann
> Axel-Naumann on Jun 29 Owner
> 
> That's a change in behavior (version might have been set to fClassVersion after your change but not before); could you document it in the commit log if it was intentional?
> @amadio
> amadio on Jun 29 • edited Member
> 
> This is intentional. When the requested version is 0, it really means the current version (which supposedly isn't itself equal to 0), so this let's the detection work properly and return the version already available (that's the point of this check). However, before the check would always fail, because sinfo->GetClassVersion() never returns 0 (at least that's how I interpreted it, correct me if I'm wrong).
> @Axel-Naumann
> Axel-Naumann on Jun 29 Owner
> 
>     (which supposedly isn't itself equal to 0)
>     ...
>     (at least that's how I interpreted it, correct me if I'm wrong)
> 
> These are assumptions; they suggest that we are probably better off waiting for Philippe with these changes?
> @amadio
> amadio on Jun 29 Member
> 
> Yes, definitely, I will not merge this before Philippe is back. I just want to give others a chance to review immediately to address any problems found before he comes back.
> @pcanal
> pcanal on Jul 3 Owner
> 
> This looks fine but really hinge on whether the line number 4366 (in new file) is thread safe or not.
> @amadio
> amadio on Jul 4 Member
> 
> That line is apparently not thread-safe, due to false sharing (lack of alignment of fClassVersion), but real problems depend on if there is any pattern where modifications to the class version can actually happen when this function is called. As commented above, we may need to go back to taking the lock despite the performance penalty. On the other hand, we need a better long term solution for our thread-safety issues. On one side we need a test suite to catch thread-safety problems like these, and on the other, we need a better way than taking locks and using atomics.
>

Comment (pcanal) [reply core/meta/src/TClass.cxx]: Note that fClassVersion is not (yet?) an atomic.  Note the implementation of TClass::GetClassVersion
Note the existence of TClass::SetClassVersion which is only use in ROOT::ResetClassVersion which relies on the atomic fVersionUsed to allow or disallow the change.   This feature is usually triggered by the use of the macro RootClassVersion (so the SetClassVersion is usually executed at library load time).

And a cursory look a the code mentions seems to indicates that the existing situation is already not really thread-safe ... (missing lock taking in the above mention function, missing setting of fVersionUsed in direct use of fClassVersion in TClass.cxx and/or fClassVersion not being atomic ....)

File: core/meta/src/TClass.cxx (hunk 3 above)
Comment (pcanal) [reply core/meta/src/TClass.cxx]: Anyway, I agree with Axel that this should be a separate commit ;)  Thanks.

File: core/meta/inc/TClass.h (hunk 1 above)
Comment (amadio) [reply core/meta/inc/TClass.h]: I think that we need to come to a point where clang-format's output *is* the coding standard, or these discussions about coding style will never end. Please adjust the configuration so that it does the right thing, then I can rebase, reformat, and push again.

Comment (pcanal) [reply core/meta/inc/TClass.h]: > Please adjust the configuration 

I can't (in good faith) ...  It just got adjusted the other way ...

> we need to come to a point where clang-format's output

As far as I can tell, this is waiting on clang-format to become 'better' (and have more features). i.e. we can not get it to do a good enough job yet.  (on a recent PR, 80% of clang-format's proposal were pessimitation of the layout).

File: core/cont/src/TList.cxx [hunk 5]
```diff
@@ -538,17 +538,20 @@ void TList::DeleteLink(TObjLink *lnk)
 
 TObject *TList::FindObject(const char *name) const
 {
+   if (!name)
```
[the remark this answers is not in the document]
Comment (amadio) [reply core/cont/src/TList.cxx]: I understand your argument, but technically, the thread is not really accessing the data from TList before the `R__COLLECTION_READ_GUARD`, so this change does not affect your ability to trace bugs. I made the change thinking about the possibility that in the future this will not be a noop.

File: core/cont/src/TList.cxx [hunk 6]
```diff
@@ -696,49 +699,47 @@ TObjLink *TList::NewOptLink(TObject *obj, Option_t *opt, TObjLink *prev)
...
-               lnk->Next()->fPrev = lnk->Prev();
-               DeleteLink(lnk);
-            }
-            fSize--;
-            fCache = 0;
-            Changed();
-         } else
-            ob->RecursiveRemove(obj);
+      // object found, remove it from the list
+      R__COLLECTION_WRITE_GUARD();
```
[the remark this answers is not in the document]
Comment (amadio) [reply core/cont/src/TList.cxx]: Same here, if in the future this becomes a real lock, it's better to only really lock when actually changing the structure than during the entire traversal of it, otherwise there will be a lot of contention for the list's first element. If performance isn't great now, it will become terrible later...

File: core/cont/src/TList.cxx (hunk 5 above)
[the remark this answers is not in the document]
Comment (pcanal) [reply core/cont/src/TList.cxx]: However, TList is known to not be thread safe and using it from several threads (without an external lock) is always a mistake no matter how far it goes.  Moving the guard to later in the function actually only reduce the effectiveness of the collision detection.

I also have no intention of making the class named TList to be thread safe (we will always need a container that does not pay the heavy price of locking for the many case where the container will not be shared).  

There might be a derived (and/or sibling) class name TListWIthLock or TListWIthRWLock or any better name.

>  I made the change thinking about the possibility that in the future this will not be a noop.

Not sure what you meant (I suspect you meant 'when it becomes a lock) since it is not a noop if you set the #define the macro R__CHECK_COLLECTION_MULTI_ACCESS.

File: core/meta/src/TClass.cxx (hunk 4 above)
Comment (amadio) [reply core/meta/src/TClass.cxx]: Indeed, we need a better solution for this kind of problem, and I suggested using [RCU](http://www.rdrop.com/~paulmck/RCU) before, which should be much better in terms of performance than using locks.

File: core/cont/src/TList.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (pcanal) [reply core/cont/src/TList.cxx]: Ah.  It will not become a lock in the future.  The semantic is different from that of a lock.  It is a collision detection mechanism, you do want to have it 'held' by the function as long as possible to increase the likelyhood of overlaps when there is one.

[the remark this answers is not in the document]
Comment (amadio) [reply core/cont/src/TList.cxx]: Alright, I will put them back where they were then. However, I'd say it's better to have lower performance than crashes... We ought to make things thread-safe in general. If it's just for lists, it shouldn't be too hard to do it while [keeping most of the performance](https://youtu.be/1Q-RH2tiyt0?t=25m32s).

[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/cont/src/TList.cxx]: But then we should also re-implement `vector` to be thread safe. And `TObject` - all its non-`const` functions must be callable from different threads. And `int` itself should really become thread-safe, after all "it's better to have lower performance than crashes..."! Or maybe not? ;-)

ROOT's collections, just like STL collections, are non-thread-safe. That by itself is not a problem, as long as users know what they get. These collections have a different role than the role you want to assign to them.

[the remark this answers is not in the document]
Comment (amadio) [reply core/cont/src/TList.cxx]: You really love distorting what people say. If a user runs a simple script with ROOT with MT enabled and it crashes on the list of cleanups, whose fault is it? The user is not using TList, yet ROOT crashes. We have to fix that. It's not acceptable.

[the remark this answers is not in the document]
Comment (amadio) [reply core/cont/src/TList.cxx]: If users use TList, etc, and shoot themselves on the foot, that's obviously another story. But ROOT cannot shoot their foot for them.

[the remark this answers is not in the document]
Comment (pcanal) [reply core/cont/src/TList.cxx]: >  We have to fix that.

Yes. Indeed and for the list of cleanups we know what we are up against.

The question those guard are intending on answering is "is there *any* TList anywhere in the ROOT source code that is inadvertently/unexpectedly used from multiple thread".

I.e. after we solve the list of cleanups problems and we still get random crash, this tool will let you know whether this is because of a TList or not ... :)

[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/cont/src/TList.cxx]: The approach so far was mostly to duplicate these collections per thread, i.e. have thread-local objects (e.g. `gDirectory`).

We have to fix crashes in the cleanup, absolutely - I just hope we can find a cheaper way than "eagerly" locking / RCUing all collections.

[the remark this answers is not in the document]
Comment (amadio) [reply core/cont/src/TList.cxx]: Now we are talking the same language :-) This is totally about the list of cleanups, which happens to be using TList and maybe also THashList. We wil either have to move away from those containers, or use them with some sort of synchronization mechanism. I'm not aware of any other way to solve the problem. Since Philippe mentioned adding a read/write lock, I thought the solution would be the latter. However, my preferred solution is to wrap the removals into a function that delegates deletion to somebody else without blocking. For example, you could have `ROOT::DestroyObject()` which puts the object's ID on a queue of transactions that is processed by a garbage collection thread, which is woken up once there's enough stuff to be destroyed.

[the remark this answers is not in the document]
Comment (pcanal) [reply core/cont/src/TList.cxx]: > Since Philippe mentioned adding a read/write lock, I thought the solution would be the latter. 

Yes, I will be adding locks ... to new classes (THashListIWithRWLocks).

> However, my preferred solution ...

This solution has some merit and some thing to be discuss (on whether it accomplish the goals/requirement or not) ... but even before that, it is a challenging option to consider as it would require user code change ... and if the user is changing their code, I would rather have them move to v7.

**Comment by amadio** [comment]
Alright, I believe I addressed all your comments. Please take another look at this new version and let me know what you think. Cheers,

**Comment by amadio** [comment]
Note, I only slightly edited commit messages, no need to retest.

**Comment by amadio** [comment]
I believe failures were due to pushing while the build was running. Everything passed before and I only changed the commit message with latest push. In any case, I will monitor the incrementals and revert if any more problems show up.
